packages/zilliqa-js-subscriptions/src/ws.ts

```diff
@@ -0,0 +1,197 @@
+//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//   GNU General Public License for more details.
+//
+//   You should have received a copy of the GNU General Public License
+//   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import mitt from 'mitt';
+import { w3cwebsocket as W3CWebsocket } from 'websocket';
+import {
+  EventType,
+  NewBlockQuery,
+  NewEventQuery,
+  SocketConnect,
+  SocketState,
+  Unsubscribe,
+} from './types';
+
+export class WebSocketProvider {
+  public static NewWebSocket(
+    url: string,
+    options: any = {},
+  ): WebSocket | W3CWebsocket {
+    // tslint:disable-next-line: no-string-literal
+    if (typeof window !== 'undefined' && (<any>window).WebSocket) {
+      // tslint:disable-next-line: no-string-literal
+      return new WebSocket(url, options.protocol);
+    } else {
+      const headers = options.headers || {};
+      const urlObject = new URL(url);
+
+      if (!headers.authorization && urlObject.username && urlObject.password) {
+        const authToken = Buffer.from(
+          `${urlObject.username}:${urlObject.password}`,
+        ).toString('base64');
+        headers.authorization = `Basic ${authToken}`;
+      }
+
+      return new W3CWebsocket(
+        url,
+        options.protocol,
+        undefined,
+        headers,
+        undefined,
+        options.clientConfig,
+      );
+    }
+  }
+
+  url: string;
+  emitter: mitt.Emitter;
+  handlers: any = {};
+  websocket: WebSocket | W3CWebsocket;
+  // use concrete object instead of `any`
+  subscriptions: any;
+
+  // basically, options is a collection of metadata things like protocol or headers
+  constructor(url: string, options: any = {}) {
+    this.url = url;
+    this.emitter = new mitt(this.handlers);
+    this.websocket = WebSocketProvider.NewWebSocket(url, options);
+    this.subscriptions = {};
+    this.websocket.onopen = this.onConnect.bind(this);
+    this.websocket.onclose = this.onClose.bind(this);
+    this.websocket.onmessage = this.onMessage.bind(this);
+  }
+
+  onClose(event: any) {
+    this.emitter.emit(SocketConnect.CLOSE, event);
+    if (this.websocket.CONNECTING) {
+      this.websocket.close();
+    }
+    return;
+  }
+
+  onError(event: any) {
+    this.emitter.emit(SocketConnect.ERROR, event);
+    if (this.websocket.CONNECTING) {
+      this.websocket.close();
+    }
+    return;
+  }
+
+  async onConnect() {
+    if (!this.subscriptions) {
+      this.subscriptions = {};
+    }
+    // todo handle reconnect issue
+    this.emitter.emit(SocketState.SOCKET_CONNECT);
+    this.emitter.emit(SocketConnect.CONNECT);
+  }
+
+  onMessage(msg: MessageEvent) {
+    if (msg.data) {
+      const dataObj = JSON.parse(msg.data);
+      if (dataObj.type === EventType.NOTIFICATION) {
+        this.emitter.emit(SocketState.SOCKET_MESSAGE, dataObj);
+        for (const value of dataObj.values) {
+          if (value.query === 'NewBlock') {
```

Comment:
Let's use value from EventType enum e.g.,`EventType.NEW_BLOCK`, instead of using 'NewBlock'